hiba-chk.1
.\" Copyright 2021 The HIBA Authors
.\"
.\" Use of this source code is governed by a BSD-style
.\" license that can be found in the LICENSE file or at
.\" https://developers.google.com/open-source/licenses/bsd
.TH HIBA-CHK 1 "Dec, 1 2020"
.SH NAME
hiba-chk - OpenSSH helper for host identity based authorizations.
.SH SYNOSPSIS
.B hiba-chk
.RI "[-v] [-y] [-l " "facility" "] [-g " "grl_file" "] -i " "identity_file" " -r " "role" " -p " "principal" " " "grant_file"
.br
.B hiba-chk
.RI "[-v] [-y] [-l " "facility" "] [-g " "grl_file" "] -i " "host_certificate" " -r " "role" " " "user_certificate"
.SH DESCRIPTION
.B hiba-chk
performs checks between a
.I host_identity
and a
.I user_grant.
.PP
Both can be provided as OpenSSH host and user certificates containing HIBA extensions generated by
.B hiba-gen
and attached to the certificate using
.B hiba-ca.sh
(or any other certificate authority that supports HIBA extensions).
.PP
Alternatively,
.B hiba-chk
accepts direct HIBA extensions, mostly for testing, or debugging authorizations. In this case, the
.I -p principal
option must be provided as it can't be extracted from the
.RI "" "user_certificate" "."
The certificate serial and issued time will be hardcoded, causing revocations to be all or nothing and validity check to always pass.
.PP
.B hiba-chk
also requires the
.I role
to grant access to and optionally a
.I grl_file
that contains the list of revoked grants inside certificates.
.SH OPTIONS
This program only accepts short options (single dash).
.TP
.B \-v
Turn on verbose mode. This flag can be repeated up to 3 times for increased verbosity.
.TP
.B \-y
.RI "Turn on syslog mode. Log to syslog (" "auth" " facility by default) instead of stderr."
.TP
.B \-l
.RI "Facility to use for logging when -y is specified. Defaults to " "auth" "."
.TP
.B \-g
The path to a grl_file.
.TP
.B \-i
The host identity, either as an openssh host certificate with a HIBA identity extension, or a direct HIBA identity extension.
.TP
.B \-r
The role to grant access to.
.TP
.B \-p
The principal to allow. This option is only necessary when not using certificates.
.SH EXIT STATUS
.B hiba-chk
display the dynamically generated
.I authorized_users
file to stdout, as expected by
.B sshd
(nothing will be printed if access is denied). It will terminate with an exit code set to zero if access is granted. If the access is denied it will set the exit code to a non-zero value representing the authorization error:
.TP 24
.B HIBA_CHECK_NOKEY
(40) One of the key from the HIBA grant cannot be found in the host identity.
.TP
.B HIBA_CHECK_BADVERSION
(41) The HIBA grant and HIBA identity versions are incompatible.
.TP
.B HIBA_CHECK_EXPIRED
(42) The HIBA grant is expired.
.TP
.B HIBA_CHECK_REVOKED
(43) The HIBA grant was revoked.
.TP
.B HIBA_CHECK_NOGRL
(44) The
.I grl_file
is specified but cannot be found.
.TP
.B HIBA_CHECK_BADHOSTNAME
(45) The HIBA grant references another hostname.
.TP
.B HIBA_CHECK_BADROLE
(46) The HIBA grant doesn't allow access as current requested role.
.TP
.B HIBA_CHECK_NOGRANTS
(47) The user certificate doesn't contain any HIBA grants.
.TP
.B HIBA_CHECK_DENIED
(48) The HIBA grant contains a key/pair not matching the host identity.
.SH EXAMPLE
.RS 4
.nf
$ hiba-gen -d -f host.hiba
identity@hibassh.dev (v1):
 [0] domain = 'google.com'
 [1] owner = 'hiba'

$ hiba-gen -d -f user.hiba
grant@hibassh.dev (v1):
 [0] domain = 'google.com'
 [1] role = 'user'

# Test access denied
$ hiba-chk -i host.hiba -r root -p principal user.hiba
$ echo $?
46

# Test access granted
$ hiba-chk -i host.hiba -r user -p principal user.hiba
principal
$ echo $?
0
.fi
.SH SEE ALSO
.BR hiba-ca.sh (1),
.BR hiba-gen (1),
.BR hiba-grl (1),
.BR sshd_config (1),
.BR ssh-keygen (1)


hiba-chk.c
/*
 * Copyright 2021 The HIBA Authors
 *
 * Use of this source code is governed by a BSD-style
 * license that can be found in the LICENSE file or at
 * https://developers.google.com/open-source/licenses/bsd
 */
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>

#define HIBA_INTERNAL

#include "log.h"
#include "hiba.h"
#include "misc.h"
#include "openbsd-compat/bsd-misc.h"
#include "util.h"

#define USER_DEFAULT_CERT_SERIAL (42 + 1)
#define HOST_DEFAULT_CERT_SERIAL (42 - 1)

void
usage() {
	fprintf(stderr, "usage:\n");
	fprintf(stderr, "  With certificates\n");
	fprintf(stderr, "    hiba-chk -i <identity> -r <role> -g <grl_path> certificate\n");
	fprintf(stderr, "  With grants\n");
	fprintf(stderr, "    hiba-chk -i <identity> -r <role> -g <grl_path> -p principal grant\n");
	exit(1);
}

int
check_access(const struct hibaenv *env, const struct hibacert *cert, const char *role) {
	int i;
	int ret;
	int len;
	int verdict = HIBA_CHECK_NOGRANTS;
	struct hibaext **grants;

	if ((ret = hibacert_hibaexts(cert, &grants, &len)) < 0)
		fatal("check_access: can't get grants from certificate: %s", hiba_err(ret));

	debug2("Testing %d grants", len);
	for (i = 0; i < len; ++i) {
		verbose("check_access: checking grant %d.", i);
		if ((ret = hibachk_authorize(env, grants[i], i, role)) == HIBA_OK) {
			verdict = HIBA_OK;
			hibachk_authorized_users(env, cert, i, stdout);
		} else {
			if (verdict != HIBA_OK)
				verdict = ret;
			verbose("check_access: denied: %s", hiba_err(ret));
		}
	}

	return verdict;
}

int
main(int argc, char **argv) {
	extern int optind;
	extern char *optarg;
	
	int opt;
	int ret;
	int debug_flag = 0;
	int use_syslog = 0;
	int log_level = SYSLOG_LEVEL_INFO;
	SyslogFacility log_facility = SYSLOG_FACILITY_AUTH;
	int grl_mmapped = 0;
	u_int64_t grl_size = 0;
	unsigned char *grl_data = NULL;
	char *grl_file = NULL;
	char *principal = NULL;
	char *identity_file = NULL;
	char *role = NULL;
	struct hibacert *host;
	struct hibacert *user;
	struct hibaext *identity;
	struct hibaext *grant;
	char *__progname = ssh_get_progname(argv[0]);

	/* Ensure that fds 0, 1 and 2 are open or directed to /dev/null. */
	sanitise_stdfd();

	while ((opt = getopt(argc, argv, "yvr:i:g:p:l:")) != -1) {
		switch (opt) {
		case 'y':
			use_syslog = 1;
			break;
		case 'v':
			if (!debug_flag) {
				debug_flag = 1;
				log_level = SYSLOG_LEVEL_DEBUG1;
			} else if (log_level < SYSLOG_LEVEL_DEBUG3) {
				log_level++;
			} else {
				fatal("Too high debugging level.");
			}
			break;
		case 'r':
			role = optarg;
			break;
		case 'g':
			grl_file = optarg;
			break;
		case 'i':
			identity_file = optarg;
			break;
		case 'p':
			principal = optarg;
			break;
		case 'l':
			log_facility = log_facility_number(optarg);
			if (log_facility == SYSLOG_FACILITY_NOT_SET)
				error("Invalid log facility \"%s\"", optarg);
			break;
		case '?':
		default:
			usage();
		}
	}

	argv += optind;
	argc -= optind;

	log_init("hiba-chk", log_level, log_facility, !use_syslog);

	if (debug_flag)
		debug2("%s: starting in debug mode", __progname);

	if (argc < 1) {
		error("%s: missing certificate/grant", __progname);
		usage();
	} else if (argc > 1) {
		error("%s: extra argument in command line", __progname);
		usage();
	}
	if (role == NULL)
		fatal("%s: missing role", __progname);
	if (identity_file == NULL)
		fatal("%s: missing host identity ", __progname);

	
	decode_file(identity_file, &host, &identity);
	decode_file(argv[0], &user, &grant);
	open_grl(grl_file, &grl_data, &grl_size, &grl_mmapped);

	{
		struct sshbuf *grl_blob = NULL;
		struct hibaenv *env;
		struct hibagrl *grl = NULL;

		if (host == NULL) {
			host = hibacert_new();
			/* If we got a HIBA identity extension with no SSH
			 * certificate, we must emulate the certificate serial
			 * and issued time. We do this with the following
			 * values: hardcoded certificate serial (causes
			 * revocations to be all or nothing), and issued on the
			 * UNIX epoch. */
			if ((ret = hibacert_from_ext(host, identity, principal, 0, HOST_DEFAULT_CERT_SERIAL)) < 0)
				fatal("%s: creating host certificate from extension: %s", __progname, hiba_err(ret));
			identity = NULL;
		}
		if (user == NULL) {
			user = hibacert_new();
			/* If we got a HIBA grant extension with no SSH
			 * certificate, we must emulate the certificate serial
			 * and issued time. We do this with the following
			 * values: hardcoded certificate serial (causes
			 * revocations to be all or nothing), and issued now
			 * (causes validity to always pass). */
			if ((ret = hibacert_from_ext(user, grant, principal, time(NULL), USER_DEFAULT_CERT_SERIAL)) < 0)
				fatal("%s: creating user certificate from extension: %s", __progname, hiba_err(ret));
			grant = NULL;
		}

		if (grl_file != NULL) {
			/* If a GRL file was specified but the content is not
                         * found, we fail closed. */
			if (grl_data == NULL) {
				fatal("%s:  cannot read GRL file %s", __progname, grl_file);
			}

			grl = hibagrl_new();
			grl_blob = sshbuf_from(grl_data, grl_size);
			if ((ret = hibagrl_decode(grl, grl_blob)) < 0) {
				fatal("%s:  cannot decode GRL file %s: %s", __progname, grl_file, hiba_err(ret));
			}
		}

		env = hibaenv_from_host(host, user, grl);
		ret = check_access(env, user, role);
		hibaenv_free(env);
		hibagrl_free(grl);
		sshbuf_free(grl_blob);
	}

	close_grl(grl_data, grl_size, grl_mmapped);
	hibacert_free(host);
	hibacert_free(user);
	hibaext_free(identity);
	hibaext_free(grant);

	free(__progname);

	return -ret;
}
